test/clj/game/cards/upgrades_test.clj

```diff
@@ -3336,3 +3336,243 @@
+    (click-prompt state :runner "0")
+    (is (= 1 (:credit (get-corp))) "After using all credits to boost trace, Corp gains 1c thanks to Traffic Analyzer")
+    ))
+
+(deftest defense-construct
+  (do-game
+    (new-game {:corp {:hand ["Defense Construct" (qty "PAD Campaign" 3)]}})
+    ;; Play 3 PAD Campaigns on the same server so 2 of them are trashed facedown
+    (play-from-hand state :corp "Defense Construct" "Archives")
+    (play-from-hand state :corp "PAD Campaign" "HQ")
+    (play-from-hand state :corp "PAD Campaign" "HQ")
+    (click-prompt state :corp "OK")
+    (take-credits state :runner)
+    (play-from-hand state :corp "PAD Campaign" "HQ")
+    (click-prompt state :corp "OK")
+    ;; Advance Defense Construct twice to recover both PAD Campaigns from discard
+    (let [dc (get-content state :archives 0)
+          advance-tokens 2]
+      (rez state :corp dc)
+      (advance state dc advance-tokens)
+      (take-credits state :corp)
+      (run-on state :archives)
+      (card-ability state :corp dc 0)
+      (click-card state :corp (get (:discard (get-corp)) 0))
+      (click-card state :corp (get (:discard (get-corp)) 1))
+      (is (= advance-tokens (count (filter #(= (:title %) "PAD Campaign") (:hand (get-corp))))) "2 advance tokens of DC so 2 PAD Campaign back to corp")
+      (is (= "Defense Construct" (:title (first (:discard (get-corp)))))))) "Defense Construct is trashed after using it")
+
+(deftest k-p-lynn
+  (do-game
+    (new-game {:corp {:hand ["K. P. Lynn"]}})
+    (play-from-hand state :corp "K. P. Lynn" "New remote")
+    ))
+
+(deftest fractal-threat-matrix
+  (do-game
+    (new-game {:corp {:hand ["Fractal Threat Matrix" "Najja 1.0"]}
+               :runner {:deck [(qty "Acacia" 7)]}})
+    (play-from-hand state :corp "Fractal Threat Matrix" "New remote")
+    (play-from-hand state :corp "Najja 1.0" "Remote 1")
+    (take-credits state :corp)
+    (is (= 2 (count (:deck (get-runner)))))
+    (run-on state :remote1)
+    (let [najja (get-ice state :remote1 0)
+          ftm (get-content state :remote1 0)]
+      (rez state :corp najja)
+      (rez state :corp ftm)
+      (run-continue state)
+      (card-side-ability state :runner najja 0)
+      (click-prompt state :runner "End the run")
+      (card-side-ability state :runner najja 0)
+      (click-prompt state :runner "End the run")
+      ;; All subroutines broken. Manually trigger Fractal Thread Matrix
+      (card-ability state :corp ftm 0)
+      (is (zero? (count (:deck (get-runner)))))
+      (is (= 2 (count (:discard (get-runner)))))
+      )))
+
+(deftest shell-corporation
+  (do-game
+    (new-game {:corp {:hand ["Shell Corporation"]}})
+    (play-from-hand state :corp "Shell Corporation" "New remote")
+    (let [sc (get-content state :remote1 0)]
+      (rez state :corp sc)
+      (card-ability state :corp (refresh sc) 0)
+      (card-ability state :corp (refresh sc) 0)             ;;Retrigger ability, no extra credits are given
```

Comment:
Comment spacing

Reply:
This line doesn't test anything because it doesn't make any assertions (typically done with `is`). Could try writing something like:
```clojure
(changes-val-macro 0 (get-counters (refresh sc) :credit) "No additional credits are placed on Shell Corp")
```